sqlstats: move sql stats writing to be async

xinhaoz opened the pull request:
This commit removes the recording of sql stats at the end of statement and
transaction execution. The per-node SQLStats container has been added as a
sink to the async sql stats ingester. Instead of having sslocal.StatsCollector
record sql stats at the end of execution, it will now record the stats
asynchronously as part of event processing in sslocal.SQLStatsIngester's
ingest routine.

When creating the execution statistics structs, we now record the application
name from the application container currently that is currenlty set in
sslocal.StatsCollector. This is because the conn exec's app name may have been
mutated if the current statement is a `set application_name` statement but we
still want to record that statement in the previous application.

Epic: none
Closes: https://github.com/cockroachdb/cockroach/issues/141024

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143855)
<!-- Reviewable:end -->

xinhaoz committed:
sqlstats: move sql stats writing to be async

This commit removes the recording of sql stats at the end of statement and
transaction execution. The per-node SQLStats container has been added as a
sink to the async sql stats ingester. Instead of having sslocal.StatsCollector
record sql stats at the end of execution, it will now record the stats
asynchronously as part of event processing in sslocal.SQLStatsIngester's
ingest routine.

When creating the execution statistics structs, we now record the application
name from the application container currently that is currenlty set in
sslocal.StatsCollector. This is because the conn exec's app name may have been
mutated if the current statement is a `set application_name` statement but we
still want to record that statement in the previous application.

At the moment this change does not have measurable impact to the
sysbench microbenchmark:
```
❯ benchdiff ./pkg/sql/tests --old 0e9251db -b -r 'Sysbench/SQL/3node/oltp_read_write' -d 1000x -c 10 --preview
old:  0e9251d Merge #150535 #150771 #150947 #150949
new:  e0f30da sqlstats: rename Ingest* functions on sql stats in
args: benchdiff "./pkg/sql/tests" "--old" "0e9251db" "-b" "-r" "Sysbench/SQL/3node/oltp_read_write" "-d" "1000x" "-c" "10" "--preview"

name                                           old time/op    new time/op    delta
Sysbench/SQL/3node/oltp_read_write-10            4.99ms ± 1%    4.97ms ± 2%   ~     (p=0.481 n=8+9)
ParallelSysbench/SQL/3node/oltp_read_write-10    1.24ms ± 3%    1.24ms ± 2%   ~     (p=0.190 n=10+10)

name                                           old errs/op    new errs/op    delta
Sysbench/SQL/3node/oltp_read_write-10              0.00           0.00        ~     (all equal)
ParallelSysbench/SQL/3node/oltp_read_write-10     0.00 ±100%     0.00 ±117%   ~     (p=1.000 n=10+10)

name                                           old alloc/op   new alloc/op   delta
Sysbench/SQL/3node/oltp_read_write-10            1.92MB ± 1%    1.92MB ± 0%   ~     (p=0.780 n=10+9)
ParallelSysbench/SQL/3node/oltp_read_write-10    1.80MB ± 1%    1.82MB ± 4%   ~     (p=0.696 n=8+10)

name                                           old allocs/op  new allocs/op  delta
Sysbench/SQL/3node/oltp_read_write-10             9.26k ± 0%     9.25k ± 0%   ~     (p=0.561 n=10+9)
ParallelSysbench/SQL/3node/oltp_read_write-10     8.09k ± 1%     8.11k ± 1%   ~     (p=0.289 n=10+10)
```

Epic: none
Closes: #141024

Release note: None

xinhaoz committed:
sqlstats: rename Ingest* functions on sql stats ingester to Buffer*

This naming more accurately reflects what these functions do with the provided
struct.

Epic: none

Release note: None

dhartunian commented:
I think the Extended CI is just timing out because the SQL Stats tests specifically are really slow.

dhartunian committed:
sqlstats: add metrics for ingester health

This patch adds two new metrics to the SQL Stats Ingester, which is
being relied on for processing all SQL stats, not just insights. One
metric provides a view of throughput but counting processed requests,
another provides a view into the queue size which should grow and
shrink periodically as flushing is triggered.

There isn't a natural place to track errors in the Ingester as the
methods it calls to process, do not return `err`.

Part of: #141024
Epic: None
Release note: None
